be/src/exec/spill/spiller.cpp

```diff
@@ -0,0 +1,230 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "exec/spill/spiller.h"
+
+#include <butil/iobuf.h>
+#include <fmt/core.h>
+#include <glog/logging.h>
+
+#include <cstdint>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <utility>
+
+#include "column/chunk.h"
+#include "common/status.h"
+#include "common/statusor.h"
+#include "exec/sort_exec_exprs.h"
+#include "exec/spill/mem_table.h"
+#include "exec/spill/spilled_stream.h"
+#include "exec/spill/spiller_path_provider.h"
+#include "gutil/port.h"
+#include "runtime/runtime_state.h"
+#include "serde/column_array_serde.h"
+
+namespace starrocks {
+SpillProcessMetrics::SpillProcessMetrics(RuntimeProfile* profile) {
+    spill_timer = ADD_TIMER(profile, "SpillTime");
+    spill_rows = ADD_COUNTER(profile, "SpilledRows", TUnit::UNIT);
+    flush_timer = ADD_TIMER(profile, "SpillFlushTimer");
+    write_io_timer = ADD_TIMER(profile, "SpillWriteIOTimer");
+    restore_rows = ADD_COUNTER(profile, "SpillRestoreRows", TUnit::UNIT);
+    restore_timer = ADD_TIMER(profile, "SpillRestoreTimer");
+}
+// Not thread safe
+class ColumnSpillFormater : public SpillFormater {
+public:
+    ColumnSpillFormater(ChunkBuilder chunk_builder) : _chunk_builder(std::move(chunk_builder)) {}
+    Status spill_as_fmt(SpillFormatContext& context, std::unique_ptr<WritableFile>& writable,
+                        const ChunkPtr& chunk) const noexcept override;
+    StatusOr<ChunkUniquePtr> restore_from_fmt(SpillFormatContext& context,
+                                              std::unique_ptr<RawInputStreamWrapper>& readable) const override;
+    Status flush(std::unique_ptr<WritableFile>& writable) const override;
+
+private:
+    size_t _spill_size(const ChunkPtr& chunk) const;
+    ChunkBuilder _chunk_builder;
+};
+
+size_t ColumnSpillFormater::_spill_size(const ChunkPtr& chunk) const {
+    size_t serialize_sz = 0;
+    for (const auto& column : chunk->columns()) {
+        serialize_sz += serde::ColumnArraySerde::max_serialized_size(*column);
+    }
+    return serialize_sz + sizeof(serialize_sz);
+}
+
+Status ColumnSpillFormater::spill_as_fmt(SpillFormatContext& context, std::unique_ptr<WritableFile>& writable,
+                                         const ChunkPtr& chunk) const noexcept {
+    size_t serialize_sz = _spill_size(chunk);
+    context.io_buffer.resize(serialize_sz);
+    DCHECK_GT(serialize_sz, 4);
+
+    auto* buff = reinterpret_cast<uint8_t*>(context.io_buffer.data());
+    UNALIGNED_STORE64(buff, serialize_sz);
+    buff += sizeof(serialize_sz);
+
+    for (const auto& column : chunk->columns()) {
+        buff = serde::ColumnArraySerde::serialize(*column, buff);
```

Comment:
as far, should compression and encodings be taken into consideration to reduce size of spill data?
it seems that WriterFile can be decorated by compression.

Reply:
it will be done in other PRs